File: +saivdr/+dcnn/nsoltBlockDct3dLayer.m
classdef nsoltBlockDct3dLayer < nnet.layer.Layer
    %NSOLTBLOCKDCT3DLAYER
    %
    %   ベクトル配列をブロック配列を入力: 
    %      (Stride(1)xnRows) x (Stride(2)xnCols) x (Stride(3)xnLays) x nComponents x nSamples
    %
    %   コンポーネント別に出力(nComponents=1):
    %      nDecs x nRows x nCols x nSamples
    %
    % Requirements: MATLAB R2020a
    %
    % Copyright (c) 2020-2021, Shogo MURAMATSU
    %
    % All rights reserved.
    %
    % Contact address: Shogo MURAMATSU,
    %                Faculty of Engineering, Niigata University,
    %                8050 2-no-cho Ikarashi, Nishi-ku,
    %                Niigata, 950-2181, JAPAN
    %
    % http://msiplab.eng.niigata-u.ac.jp/
    
    properties
        % (Optional) Layer properties.
        DecimationFactor
        
        % Layer properties go here.
    end
    
    properties (Access = private)
        Cvhd
    end
    
    methods
        
        function layer = nsoltBlockDct3dLayer(varargin)
            % (Optional) Create a myLayer.
            % This function must have the same name as the class.
            import saivdr.dictionary.utility.Direction
            p = inputParser;
            addParameter(p,'DecimationFactor',[])
            addParameter(p,'Name','')
            addParameter(p,'NumberOfComponents',1)
            parse(p,varargin{:})
            
            % Layer constructor function goes here.
            layer.DecimationFactor = p.Results.DecimationFactor;
            layer.Name = p.Results.Name;
            layer.Description = "Block DCT of size " ...
                + layer.DecimationFactor(1) + "x" ...
                + layer.DecimationFactor(2) + "x" ...
                + layer.DecimationFactor(3);
            layer.Type = '';
            layer.NumOutputs = p.Results.NumberOfComponents;
            layer.NumInputs = 1;
            
            decV = layer.DecimationFactor(Direction.VERTICAL);
            decH = layer.DecimationFactor(Direction.HORIZONTAL);
            decD = layer.DecimationFactor(Direction.DEPTH);
            %
            Cv_ = dctmtx(decV);
            Ch_ = dctmtx(decH);
            Cd_ = dctmtx(decD);
            Cv_ = [ Cv_(1:2:end,:) ; Cv_(2:2:end,:) ];
            Ch_ = [ Ch_(1:2:end,:) ; Ch_(2:2:end,:) ];
            Cd_ = [ Cd_(1:2:end,:) ; Cd_(2:2:end,:) ];
            %
            Cve = Cv_(1:ceil(decV/2),:);
            Cvo = Cv_(ceil(decV/2)+1:end,:);
            Che = Ch_(1:ceil(decH/2),:);
            Cho = Ch_(ceil(decH/2)+1:end,:);
            Cde = Cd_(1:ceil(decD/2),:);
            Cdo = Cd_(ceil(decD/2)+1:end,:);
            %
            Cee = kron(Che,Cve);
            Coo = kron(Cho,Cvo);
            Coe = kron(Che,Cvo);
            Ceo = kron(Cho,Cve);
            %
            Ceee = kron(Cde,Cee);
            Ceoo = kron(Cdo,Ceo);
            Cooe = kron(Cde,Coo);
            Coeo = kron(Cdo,Coe);
            Ceeo = kron(Cdo,Cee);
            Ceoe = kron(Cde,Ceo);
            Cooo = kron(Cdo,Coo);
            Coee = kron(Cde,Coe);
            layer.Cvhd = ...
                [ Ceee; Ceoo; Cooe; Coeo; Ceeo; Ceoe; Cooo; Coee ]; % Cyxz
            
        end
        
        function varargout = predict(layer, X)
            % Forward input data through the layer at prediction time and
            % output the result.
            %
            % Inputs:
            %         layer       - Layer to forward propagate through
            %         X           - Input data
            % Outputs:
            %         Z1, ..., Zm - Outputs of layer forward function
            import saivdr.dictionary.utility.Direction
            varargout = cell(1,nargout);
            
            if isdlarray(X)
                X = stripdims(X);
            end
            
            % Layer forward function for prediction goes here.
            decFactor = layer.DecimationFactor;
            decV = decFactor(Direction.VERTICAL);
            decH = decFactor(Direction.HORIZONTAL);
            decD = decFactor(Direction.DEPTH);
            nDec = decV*decH*decD;
            %
            Cvhd_ = layer.Cvhd;
            %
            height = size(X,1);
            width = size(X,2);
            depth = size(X,3);
            nComponents = size(X,4);
            nSamples = size(X,5);
            nRows = height/decV;
            nCols = width/decH;
            nLays = depth/decD;
            %
            for iComponent = 1:nComponents
                arrayX = permute(reshape(X(:,:,:,iComponent,:),...
                    decV,nRows,decH,nCols,decD,nLays,nSamples),...
                    [1 3 5 2 4 6 7]);
                if isgpuarray(X)
                    varargout{iComponent} = pagefun(@mtimes,Cvhd_,...
                        reshape(arrayX,nDec,nRows,nCols,nLays,nSamples));
                else
                    varargout{iComponent} = reshape(...
                        Cvhd_*reshape(arrayX,nDec,[]),...
                        nDec,nRows,nCols,nLays,nSamples);
                end
            end
        end
        
        function dLdX = backward(layer, varargin)
            % (Optional) Backward propagate the derivative of the loss
            % function through the layer.
            %
            % Inputs:
            %         layer             - Layer to backward propagate through
            %         X1, ..., Xn       - Input data
            %         Z1, ..., Zm       - Outputs of layer forward function
            %         dLdZ1, ..., dLdZm - Gradients propagated from the next layers
            %         memory            - Memory value from forward function
            % Outputs:
            %         dLdX1, ..., dLdXn - Derivatives of the loss with respect to the
            %                             inputs
            %         dLdW1, ..., dLdWk - Derivatives of the loss with respect to each
            %                             learnable parameter
            import saivdr.dictionary.utility.Direction
            nComponents = layer.NumOutputs;
            
            % Layer forward function for prediction goes here.
            decFactor = layer.DecimationFactor;
            decV = decFactor(Direction.VERTICAL);
            decH = decFactor(Direction.HORIZONTAL);
            decD = decFactor(Direction.DEPTH);
            nDec = decV*decH*decD;
            Cvhd_T = layer.Cvhd.';
            %
            dLdZ = varargin{layer.NumInputs+layer.NumOutputs+1};            
            nRows = size(dLdZ,2);
            nCols = size(dLdZ,3);
            nLays = size(dLdZ,4);
            height = decV*nRows;
            width = decH*nCols;
            depth = decD*nLays;
            nSamples = size(dLdZ,5);
            dLdX = zeros(height,width,depth,nComponents,nSamples,'like',dLdZ);
            %
            for iComponent = 1:nComponents
                dLdZ = varargin{layer.NumInputs+layer.NumOutputs+iComponent};
                if isgpuarray(dLdZ)
                    arrayX = pagefun(@mtimes,Cvhd_T,dLdZ);
                else
                    arrayX = Cvhd_T*reshape(dLdZ,nDec,[]);
                end
                dLdX(:,:,:,iComponent,:) = ...
                    reshape(ipermute(reshape(arrayX,...
                    decV,decH,decD,nRows,nCols,nLays,nSamples),...
                    [1 3 5 2 4 6 7]),...
                    height,width,depth,1,nSamples);
            end
        end
    end
end

File: appendix/torch_nsolt/nsoltBlockDct3dLayer.py
import torch
import torch.nn as nn
import math
from nsoltUtility import Direction, block_dct_matrix_3d

class NsoltBlockDct3dLayer(nn.Module):
    """
    NSOLTBLOCKCDT3DLAYER
    
       ベクトル配列をブロック配列を入力:
          nSamples x nComponents x (Stride[0]xnRows) x (Stride[1]xnCols) x (Stride[2]xnLays)
    
       コンポーネント別に出力(nComponents):
          nSamples x nDecs x nRows x nCols x nLays
        
    Requirements: Python 3.7.x, PyTorch 1.7.x
    
    Copyright (c) 2020-2021, Yuya Kodama, Shogo MURAMATSU
    
    All rights reserved.
    
    Contact address: Shogo MURAMATSU,
        Faculty of Engineering, Niigata University,
        8050 2-no-cho Ikarashi, Nishi-ku,
        Niigata, 950-2181, JAPAN
    
        http://msiplab.eng.niigata-u.ac.jp/
    """

    def __init__(self,
        name='',
        decimation_factor=[],
        number_of_components=1
        ):
        super(NsoltBlockDct3dLayer, self).__init__()
        self.decimation_factor = decimation_factor
        self.name = name
        self.description = "Block DCT of size " \
            + str(self.decimation_factor[Direction.VERTICAL]) + "x" \
            + str(self.decimation_factor[Direction.HORIZONTAL]) + "x" \
            + str(self.decimation_factor[Direction.DEPTH])
        #self.type = ''
        self.num_outputs = number_of_components
        #self.num_inputs = 1

    def forward(self,X):
        nComponents = self.num_outputs
        nSamples = X.size(0)
        height = X.size(2)
        width = X.size(3)
        depth = X.size(4)
        stride = self.decimation_factor
        decV = stride[Direction.VERTICAL]
        decH = stride[Direction.HORIZONTAL]
        decD = stride[Direction.DEPTH]
        nrows = int(math.ceil(height/decV))
        ncols = int(math.ceil(width/decH))
        nlays = int(math.ceil(depth/decD))
        ndecs = decV*decH*decD

        # Block DCT matrix (the same as Cvhd in MATLAB nsoltBlockDct3dLayer)
        Cvhd = block_dct_matrix_3d(stride,dtype=X.dtype,device=X.device)
        # Split into decV x decH x decD blocks, whose voxels are arranged in
        # column-major order as in MATLAB:
        # (nSamples x nComponents x nrows x ncols x nlays) x (decD x decH x decV)
        arrayX = X.reshape(nSamples,nComponents,nrows,decV,ncols,decH,nlays,decD)\
            .permute(0,1,2,4,6,7,5,3)\
            .reshape(nSamples,nComponents,nrows,ncols,nlays,ndecs)
        # Apply the DCT: nSamples x nComponents x nrows x ncols x nlays x ndecs
        Z = arrayX @ Cvhd.T

        if nComponents<2:
            return torch.squeeze(Z,dim=1)
        else:
            return map(lambda x: torch.squeeze(x,dim=1),torch.chunk(Z,nComponents,dim=1))
